larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,470 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc
+ *
+ *  @brief  Implementation file for the reclustering algorithm class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "Managers/ClusterManager.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArPcaHelper.h"
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+
+#include "larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+ThreeDReclusteringAlgorithm::ThreeDReclusteringAlgorithm():
+    m_hitThresholdForNewPfo(0),
+    m_fomThresholdForReclustering(0.3)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ThreeDReclusteringAlgorithm::~ThreeDReclusteringAlgorithm()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::Run()
+{
+    //Only proceed if successfuly get shower pfo list, and it has at least one pfo
+    const PfoList *pShowerPfoList(nullptr);
+    if ((STATUS_CODE_SUCCESS == PandoraContentApi::GetList(*this, "ShowerParticles3D", pShowerPfoList)) && pShowerPfoList->size())
+    {
+        m_newPfosListNameAllAfterReclustering = "newShowerParticles3D";
+        std::string newPfosListNameUnchanged = "unchangedShowerParticles3D";
+        PfoList unchangedPfoList;   
+
+        //Save current pfo list name, so that this can be set as current again at the end of reclustering
+        std::string initialPfosListName;
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentListName<Pfo>(*this, initialPfosListName));
+
+        //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+        const ClusterList *pShowerClusters(nullptr);
+        PandoraContentApi::GetList(*this, "ShowerClusters3D", pShowerClusters);
+        if(!pShowerClusters) return STATUS_CODE_NOT_FOUND;
+
+        for (const Pfo *const pShowerPfo : *pShowerPfoList)
+        {
+            ClusterList clusterList3D;
+            LArPfoHelper::GetThreeDClusterList(pShowerPfo, clusterList3D);
+           
+            if(!this->PassesCutsForReclustering(pShowerPfo))
+            {
+                unchangedPfoList.push_back(pShowerPfo);
+                continue;
+            }
+            if (clusterList3D.empty())
+                continue;
```

Comment:
Are both of these if statements needed?

It looks like the first thing `PassesCutsForReclustering` does it load this cluster list in, and check if its empty, so probably don't need both, as if it is empty, we'd stop in the first if statement anyways.

We could also swap `PassesCutsForReclustering` to take `clusterList3D` as a second argument, as currently the code will populate the list, then run `PassesCutsForReclustering` and make and repopulate the same list.

Reply:
Removed the second if, thanks. I have played a bit with the content of "PassesCutsForReclustering" in the past, so this must be a residue of some previous logic.